# EXPLORING USED CARS LISTINGS ON EBAY

In [7]:
import pandas as pd
import numpy as np
autos = pd.read_csv("/content/drive/MyDrive/Dataquest/Ebay_sales/autos.csv", encoding = "Windows-1252")


In [49]:
autos.sample(15)

date_crawled  ...            last_seen
18728  2016-03-12 17:50:06  ...  2016-03-19 21:18:06
10272  2016-03-12 09:59:06  ...  2016-04-06 00:44:42
27827  2016-03-28 14:42:47  ...  2016-04-06 18:19:32
72     2016-03-14 07:55:36  ...  2016-04-07 08:17:00
24247  2016-03-21 18:50:22  ...  2016-04-07 06:46:13
7854   2016-03-29 21:54:13  ...  2016-04-06 09:17:20
32999  2016-03-29 18:50:45  ...  2016-04-06 04:45:49
40508  2016-03-24 12:37:37  ...  2016-04-01 06:18:09
21164  2016-03-29 22:37:21  ...  2016-03-31 18:17:32
6615   2016-03-15 18:52:37  ...  2016-03-15 18:52:37
31552  2016-03-19 16:56:37  ...  2016-03-20 19:45:42
19215  2016-03-28 20:55:46  ...  2016-04-07 03:45:06
13831  2016-03-10 16:54:29  ...  2016-03-12 07:16:27
36326  2016-04-05 11:57:16  ...  2016-04-05 11:57:16
7526   2016-03-21 08:56:25  ...  2016-04-06 02:45:23

[15 rows x 20 columns]

In [9]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [10]:
autos.info()
print('\n')
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled  ...             lastSeen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

In [11]:
autos.columns
copy_columns = autos.columns
print(copy_columns)



Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [12]:
autos = autos.rename({
    'dateCrawled' : 'date_crawled', 
    'name' : 'name', 
    'seller' : 'seller', 
    'offerType' : 'offer_type', 
    'price' : 'price', 
    'abtest' : 'abtest',
    'vehicleType' : 'vehicle_type', 
    'yearOfRegistration' : 'registration_year', 
    'gearbox' : 'gearbox', 
    'powerPS' : 'powerPS', 
    'model' : 'model',
    'odometer' : 'odometer', 
    'monthOfRegistration' : 'registration_month', 
    'fuelType' : 'fuel_type', 
    'brand' : 'brand',
    'notRepairedDamage' : 'unrepaired_damage', 
    'dateCreated' : 'ad_created', 
    'nrOfPictures' : 'nr_of_pictures', 
    'postalCode' : 'postal_code',
    'lastSeen' : 'last_seen'
}, axis='columns')


In [13]:
print(autos.columns)

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'powerPS', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')


In [14]:
autos.head()


date_crawled  ...            last_seen
0  2016-03-26 17:47:46  ...  2016-04-06 06:45:54
1  2016-04-04 13:38:56  ...  2016-04-06 14:45:08
2  2016-03-26 18:57:24  ...  2016-04-06 20:15:37
3  2016-03-12 16:58:10  ...  2016-03-15 03:16:28
4  2016-04-01 14:38:50  ...  2016-04-01 14:38:50

[5 rows x 20 columns]

### The column titles are now normalized

In [15]:
autos.describe()

registration_year       powerPS  ...  nr_of_pictures   postal_code
count       50000.000000  50000.000000  ...         50000.0  50000.000000
mean         2005.073280    116.355920  ...             0.0  50813.627300
std           105.712813    209.216627  ...             0.0  25779.747957
min          1000.000000      0.000000  ...             0.0   1067.000000
25%          1999.000000     70.000000  ...             0.0  30451.000000
50%          2003.000000    105.000000  ...             0.0  49577.000000
75%          2008.000000    150.000000  ...             0.0  71540.000000
max          9999.000000  17700.000000  ...             0.0  99998.000000

[8 rows x 5 columns]

In [16]:
#autos['postal_code'].value_counts()

print(autos['odometer'].dtypes)
print(autos['price'].dtypes)


object
object


In [17]:
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "")

autos["price"] = autos["price"].astype(float)

In [18]:
print(autos['price'].dtypes)

float64


In [19]:
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "")

autos["odometer"] = autos["odometer"].astype(float)

In [20]:
print(autos['odometer'].dtypes)

float64


In [21]:
autos.rename({"odometer" : "odometer_km"}, axis=1, inplace=True)

In [22]:
print(autos["odometer_km"].min())
print(autos["odometer_km"].max())
print('\n')
print(autos["price"].min())
print(autos["price"].max())

5000.0
150000.0


0.0
99999999.0


In [23]:
n = len(pd.unique(autos['odometer_km'])) 
print("No.of.unique odometer values :",  
      n)

b = len(pd.unique(autos['price'])) 
print("No.of.unique price values :",  
      n)

b = len(pd.unique(autos['registration_year'])) 
print("No.of.unique registration year values :",  
      n)

No.of.unique odometer values : 13
No.of.unique price values : 13
No.of.unique registration year values : 13


In [24]:
print(len(autos['price'].unique()))
print(len(autos['odometer_km'].unique()))
print(len(autos['registration_year'].unique()))

2357
13
97


In [25]:
autos["odometer_km"].value_counts().sort_index(ascending = False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_km, dtype: int64

In [26]:
autos["price"].value_counts().sort_index(ascending = False).head(20)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
Name: price, dtype: int64

In [27]:
autos["price"].value_counts().sort_index(ascending = True).head(40)

0.0      1421
1.0       156
2.0         3
3.0         1
5.0         2
8.0         1
9.0         1
10.0        7
11.0        2
12.0        3
13.0        2
14.0        1
15.0        2
17.0        3
18.0        1
20.0        4
25.0        5
29.0        1
30.0        7
35.0        1
40.0        6
45.0        4
47.0        1
49.0        4
50.0       49
55.0        2
59.0        1
60.0        9
65.0        5
66.0        1
70.0       10
75.0        5
79.0        1
80.0       15
89.0        1
90.0        5
99.0       19
100.0     134
110.0       3
111.0       2
Name: price, dtype: int64

In [28]:
autos = autos[autos["price"].between(1,351000)]
autos.describe()

price  registration_year  ...  nr_of_pictures   postal_code
count   48565.000000       48565.000000  ...         48565.0  48565.000000
mean     5888.935591        2004.755421  ...             0.0  50975.745207
std      9059.854754          88.643887  ...             0.0  25746.968398
min         1.000000        1000.000000  ...             0.0   1067.000000
25%      1200.000000        1999.000000  ...             0.0  30657.000000
50%      3000.000000        2004.000000  ...             0.0  49716.000000
75%      7490.000000        2008.000000  ...             0.0  71665.000000
max    350000.000000        9999.000000  ...             0.0  99998.000000

[8 rows x 7 columns]

### We removed the price datas after 350k because it's the limit where the bid started to raise abnormaly (from 350k to 999k)

In [29]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [30]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

### As registration_year indicates, some values seems abnormal, like 1000 for the earliest year, and 9999 for the oldest

In [31]:
autos["registration_year"].value_counts().sort_index(ascending=False)

9999    3
9000    1
8888    1
6200    1
5911    1
       ..
1910    5
1800    2
1111    1
1001    1
1000    1
Name: registration_year, Length: 95, dtype: int64

### The values above 2019 shoul be deleted, as welle as the values under 1910, given that the word itself occures in 1901

In [32]:
print(autos["registration_year"].value_counts(normalize=True).sort_index())



1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
          ...   
5911    0.000021
6200    0.000021
8888    0.000021
9000    0.000021
9999    0.000062
Name: registration_year, Length: 95, dtype: float64


In [33]:
autos = autos[autos["registration_year"].between(1900,2016)]

In [34]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=False).head(10)

2016    0.026135
2015    0.008397
2014    0.014203
2013    0.017202
2012    0.028063
2011    0.034768
2010    0.034040
2009    0.044665
2008    0.047450
2007    0.048778
Name: registration_year, dtype: float64

### Let's check if these values are removed by calculating their proportion: 


In [35]:
(~autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.0

#### Everything is OK
Let's check the brand column :

In [36]:
brand_repartition = (autos["brand"].value_counts() / autos["brand"].value_counts().sum())*100
print(brand_repartition)

volkswagen        21.126368
bmw               11.004477
opel              10.758124
mercedes_benz      9.646323
audi               8.656627
ford               6.989996
renault            4.714980
peugeot            2.984083
fiat               2.564212
seat               1.827296
skoda              1.640925
nissan             1.527388
mazda              1.518819
smart              1.415994
citroen            1.400998
toyota             1.270324
hyundai            1.002549
sonstige_autos     0.981127
volvo              0.914719
mini               0.876159
mitsubishi         0.822604
honda              0.784045
kia                0.706926
alfa_romeo         0.664082
porsche            0.612669
suzuki             0.593389
chevrolet          0.569825
chrysler           0.351321
dacia              0.263490
daihatsu           0.250637
jeep               0.227073
subaru             0.214220
land_rover         0.209936
saab               0.164949
jaguar             0.156381
daewoo             0

### We'll keep every brand over 3% of recurrency

In [38]:
brand_top = brand_repartition[brand_repartition > 3].index
print(brand_top)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford',
       'renault'],
      dtype='object')


In [39]:
avg_price_per_brand = {}
for brand in brand_top:
    brand_name = autos[autos["brand"] == brand]
    mean_price = brand_name["price"].mean()
    avg_price_per_brand[brand] = int(mean_price)
    
print(avg_price_per_brand)
print('\n')

a = sorted(avg_price_per_brand.items(), key=lambda x: x[1], reverse=True)    
print(a)



{'volkswagen': 5402, 'bmw': 8332, 'opel': 2975, 'mercedes_benz': 8628, 'audi': 9336, 'ford': 3749, 'renault': 2474}


[('audi', 9336), ('mercedes_benz', 8628), ('bmw', 8332), ('volkswagen', 5402), ('ford', 3749), ('opel', 2975), ('renault', 2474)]


#### We see a gap in mean prices, between the top 3 expensive brands (Mercedes, Audi, BMW), and the other 4

### Let's convert this list in a panda series : 

In [40]:
appb_series = pd.Series(avg_price_per_brand)
print(appb_series)



volkswagen       5402
bmw              8332
opel             2975
mercedes_benz    8628
audi             9336
ford             3749
renault          2474
dtype: int64


### And convert this series in a panda frame

In [41]:
mean_price_df = pd.DataFrame(appb_series, columns=['mean_price'])
print(mean_price_df)

               mean_price
volkswagen           5402
bmw                  8332
opel                 2975
mercedes_benz        8628
audi                 9336
ford                 3749
renault              2474


### Let's create a dataframe for mean mileage (odometer_km) : 
(first by creating a dictionnary for that value in the top brands)

In [42]:

#odometer_km
#mean_price = {}
brand_mean_mileage = {}

for brand in brand_top:
    brand_name = autos[autos["brand"] == brand]
    mean_mileage = brand_name["odometer_km"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)
    
print(brand_mean_mileage)
    

    

{'volkswagen': 128707, 'bmw': 132572, 'opel': 129310, 'mercedes_benz': 130788, 'audi': 129157, 'ford': 124266, 'renault': 128071}


In [46]:
mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
#mean_prices = pd.Series(mean_price_df).sort_values(ascending=False)
mean_mileage

bmw              132572
mercedes_benz    130788
opel             129310
audi             129157
volkswagen       128707
renault          128071
ford             124266
dtype: int64

In [47]:
mean_mileage_df = pd.DataFrame(mean_mileage, columns=['mean_mileage'])
print(mean_mileage_df)

               mean_mileage
bmw                  132572
mercedes_benz        130788
opel                 129310
audi                 129157
volkswagen           128707
renault              128071
ford                 124266


In [48]:
mean_mileage_df["mean_price"] = mean_price_df
mean_mileage_df

mean_mileage  mean_price
bmw                  132572        8332
mercedes_benz        130788        8628
opel                 129310        2975
audi                 129157        9336
volkswagen           128707        5402
renault              128071        2474
ford                 124266        3749

## **As a conclusion, we noticed that the most expensive brands tend to be sold at a higher mileage than the cheapest** 